packages/vsx-registry/src/common/vsx-registry-api.ts

```diff
@@ -131,31 +131,56 @@ export class VSXRegistryAPI {
 
     /**
      * Get the latest compatible extension version.
+     * - a builtin extension is fetched based on the extension version which matches the API.
      * - an extension satisfies compatibility if its `engines.vscode` version is supported.
      * @param id the extension id.
      *
      * @returns the data for the latest compatible extension version if available, else `undefined`.
      */
     async getLatestCompatibleExtensionVersion(id: string): Promise<VSXExtensionRaw | undefined> {
         const extensions = await this.getAllVersions(id);
-        for (let i = 0; i < extensions.length; i++) {
-            const extension: VSXExtensionRaw = extensions[i];
-            if (extension.engines && this.isEngineSupported(extension.engines.vscode)) {
-                return extension;
+        if (extensions.length === 0) {
+            return undefined;
+        }
+
+        const namespace = extensions[0].namespace.toLowerCase();
+        if (this.isBuiltinNamespace(namespace)) {
+            const apiVersion = this.apiVersionProvider.getApiVersion();
+            for (const extension of extensions) {
+                if (this.isVersionLTE(extension.version, apiVersion)) {
+                    return extension;
+                }
+            }
+            console.log(`Skipping: built-in extension "${id}" at version "${apiVersion}" does not exist.`);
+        } else {
+            for (const extension of extensions) {
+                if (this.isEngineSupported(extension.engines?.vscode)) {
+                    return extension;
+                }
             }
         }
     }
 
     /**
      * Get the latest compatible version of an extension.
-     * @param versions the `allVersions` property.
+     * @param entry the extension search entry.
      *
      * @returns the latest compatible version of an extension if it exists, else `undefined`.
      */
-    getLatestCompatibleVersion(versions: VSXAllVersions[]): VSXAllVersions | undefined {
-        for (const version of versions) {
-            if (this.isEngineSupported(version.engines?.vscode)) {
-                return version;
+    getLatestCompatibleVersion(entry: VSXSearchEntry): VSXAllVersions | undefined {
```

Comment:
This code is duplicated from previous method and can be abstracted to a common method

Reply:
Unfortunately it might be a bit difficult to cleanly refactor the code where we handle the `builtins` for both.
Each method takes a different parameter and is expected to return a different type.

Reply:
OK, It's fine to leave it as is then!